### PR TITLE
[dalgona] Add single Op tests directory

### PR DESCRIPTION
**Pull request by jinevening**
This adds single Op tests directory.

ONE-DCO-1.0-Signed-off-by: Hyukjin Jeong <hj1.jeong@samsung.com>

---
Related to: https://github.com/Samsung/ONE/pull/3501

### CODE REVIEW

File: compiler/dalgona/tests/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,72 @@
+if(NOT ENABLE_TEST)
+  return()
+endif(NOT ENABLE_TEST)
+
+unset(DALGONA_SINGLE_OP_TEST)
+
+nnas_find_package(FlatBuffers EXACT 2.0 QUIET)
+if(NOT FlatBuffers_FOUND)
+  message(STATUS "Build dalgona: FAILED (missing FlatBuffers)")
```
Comment (seanshpark) [inline compiler/dalgona/tests/CMakeLists.txt:9]: maybe "Test dalgona:" ?

File: compiler/dalgona/tests/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+endif(NOT FlatBuffers_FOUND)
+
+macro(singleOpTest NAME)
+  list(APPEND DALGONA_SINGLE_OP_TEST ${NAME})
+endmacro(singleOpTest)
+
+# Read "test.lst"
+include("test.lst")
+# Read "test.local.lst" if exists
+include("test.local.lst" OPTIONAL)
```
Comment (seanshpark) [inline compiler/dalgona/tests/CMakeLists.txt:20]: I think you need to add `.gitignore` with `/test.local.lst` in `compiler/dalgona`

File: compiler/dalgona/tests/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+###
+### Generate python interface for circle schema
+###
+set(CIRCLE_SCHEMA_PYTHON_DIR "${CMAKE_CURRENT_BINARY_DIR}/circle")
+
+add_custom_command(
+  OUTPUT ${CIRCLE_SCHEMA_PYTHON_DIR}
+  COMMAND ${CMAKE_COMMAND} -E remove_directory "${CIRCLE_SCHEMA_PYTHON_DIR}"
+  COMMAND "$<TARGET_FILE:flatbuffers::flatc>" --python
+  -o "${CMAKE_CURRENT_BINARY_DIR}" "${SCHEMA_BIN_PATH}/schema.fbs"
```
Comment (seanshpark) [inline compiler/dalgona/tests/CMakeLists.txt:54]: (optional) add some indents to alight with `"$<...` ?

File: compiler/dalgona/tests/SingleOperatorTest.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
+# Copyright 2022 Samsung Electronics Co., Ltd. All Rights Reserved.
```
Comment (seanshpark) [inline compiler/dalgona/tests/SingleOperatorTest.py:1]: (1) this file name has different convention with other scripts. either way, would be better to follow one style
(2) this has copyright notice, this too would be better to follow one style

Comment (jinevening) [reply compiler/dalgona/tests/SingleOperatorTest.py:1]: I changed the file name style (except test.lst for consistency with other tests), and copied copyright notice. PTAL

File: compiler/dalgona/tests/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+###
+### Generate python interface for circle schema
+###
+set(CIRCLE_SCHEMA_PYTHON_DIR "${CMAKE_CURRENT_BINARY_DIR}/circle")
+
+add_custom_command(
+  OUTPUT ${CIRCLE_SCHEMA_PYTHON_DIR}
+  COMMAND ${CMAKE_COMMAND} -E remove_directory "${CIRCLE_SCHEMA_PYTHON_DIR}"
+  COMMAND "$<TARGET_FILE:flatbuffers::flatc>" --python
+    -o "${CMAKE_CURRENT_BINARY_DIR}" "${SCHEMA_BIN_PATH}/schema.fbs"
```
Comment (seanshpark) [inline compiler/dalgona/tests/CMakeLists.txt:53]: (optional) just to be sure,
```suggestion
  COMMAND "$<TARGET_FILE:flatbuffers::flatc>" --python
          -o "${CMAKE_CURRENT_BINARY_DIR}" "${SCHEMA_BIN_PATH}/schema.fbs"
```

**Review by seanshpark — DISMISSED** [review]
LGTM thank you!

**Review by seanshpark — APPROVED** [review]
LGTM

File: compiler/dalgona/tests/GenH5RandomInputs.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+group.attrs['desc'] = "Input data for " + model
+
+
+def toNumpyType(circle_type):
+    if circle_type == TensorType.UINT8:
+        return np.uint8
+    if circle_type == TensorType.FLOAT32:
+        return np.float32
+    if circle_type == TensorType.INT16:
+        return np.int16
```
Comment (mhs4670go) [inline compiler/dalgona/tests/GenH5RandomInputs.py:62]: (optioanl) You can revise the logic like below.
```python
def toNumpyType(circle_type):
    NUMPY_T = [
        np.float32, np.float64, np.int8, np.int16, np.int32, np.int64, np.uint8, np.uint32, np.uint64,
        np.bool_
    ]
    circle_to_np = {getattr(circle.TensorType, np.dtype(t).name.upper()): t for t in NUMPY_T}
    return circle_to_np[circle_type]
```

Comment (seanshpark) [reply compiler/dalgona/tests/GenH5RandomInputs.py:62]: @mhs4670go 's suggestion looks simple. adding comment about `circle.TensorType` items would help understanding without searching for this type :)

File: compiler/dalgona/tests/GenH5RandomInputs.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+                0, high=256, size=tensor.ShapeAsNumpy(), dtype=np_type)
+        elif np_type == np.float32:
+            input_data = np.array(10 * np.random.random_sample(tensor.ShapeAsNumpy()) - 5,
+                                  np_type)
+        elif np_type == np.int16:
+            input_data = np.random.randint(
+                -32767, high=32768, size=tensor.ShapeAsNumpy(), dtype=np_type)
+        else:
+            raise SystemExit('Unsupported data type')
+        sample.create_dataset(str(j), data=input_data)
```
Comment (mhs4670go) [inline compiler/dalgona/tests/GenH5RandomInputs.py:83]: How about introduce random generator for readability and extensibility?

```python
class RandomDataGenerator:
  def __init__(self, shape):
    self.tensor = tensor
    self.shape = shape

  def _unsupport_types(self, dtype):
    raise RuntimeError('Unsupported data type')

  def _gen_uint8(self, dtype):
   return np.random.randint(0, high=256, size=self.shape, dtype=toNumpyType(dtype))

  def _gen_float32(self, dtype):
    return np.array(10 * np.random.random_sample(self.shape) - 5, toNumpyType(dtype))

  def gen(self, dtype):
    gen_book = dict()
    gen_book[TensorType.UINT8] = self._gen_uint8
    gen_book[TensorType.FLOAT32] = self._gen_float32

    return gen_book.get(dtype, self._unsupport_types)(dtype)

for j in range(len(inputs)):
    input_index = inputs[j]
    tensor = graph.Tensors(input_index)
    g = RandomDataGenerator(tensor.ShapeAsNumpy())
    # g.gen(tensor.Type())
```

File: compiler/dalgona/tests/GenH5RandomInputs.py (hunk 6 above)
Comment (mhs4670go) [reply compiler/dalgona/tests/GenH5RandomInputs.py:62]: > adding comment about `circle.TensorType` items would help understanding without searching for this type :)

Right. Actually, just a dictionary { CIRCLE_TYPE : NUMPY_TYPE } might be enough.

File: compiler/dalgona/tests/SingleOperatorTest.py (hunk 4 above)
Comment (mhs4670go) [inline compiler/dalgona/tests/SingleOperatorTest.py:1]: (trivial) It might need shebang.

File: compiler/dalgona/tests/GenH5RandomInputs.py (hunk 7 above)
Comment (mhs4670go) [reply compiler/dalgona/tests/GenH5RandomInputs.py:83]: (optioanl2) And, introducing test of `GenH5RandomInputs.py` would be helpful.

File: compiler/dalgona/tests/GenH5RandomInputs.py (hunk 6 above)
Comment (jinevening) [reply compiler/dalgona/tests/GenH5RandomInputs.py:62]: I've made RandomDataGenerator and directly link `circle.TensorType` with generator functions.

File: compiler/dalgona/tests/GenH5RandomInputs.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
+#!/usr/bin/env python3
```
Comment (seanshpark) [inline compiler/dalgona/tests/GenH5RandomInputs.py:1]: Q) does this file run in virtual enviroment with python3.8 installed ?

File: compiler/dalgona/tests/GenH5RandomInputs.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+import h5py as h5
+import numpy as np
+from circle.Model import Model
+import argparse
```
Comment (seanshpark) [inline compiler/dalgona/tests/GenH5RandomInputs.py:21]: (optional)
```suggestion
import h5py as h5
import numpy as np
import argparse

from circle.Model import Model
```

Comment (seanshpark) [reply compiler/dalgona/tests/GenH5RandomInputs.py:21]: Q) maybe new `compiler/dalgona` is not format checked for python files...
recent `onecc-docker` draft module didn't either...
@jinevening can you please check the format checker is working correctly?

**Review by seanshpark — APPROVED** [review]
LGTM thank you!

File: compiler/dalgona/tests/GenH5RandomInputs.py (hunk 9 above)
Comment (jinevening) [reply compiler/dalgona/tests/GenH5RandomInputs.py]: format checker works for "*.py" files. I've checked that format checker checks this file :)

**Review by mhs4670go — APPROVED** [review]
LGTM

**Pull request merged**